### PR TITLE
Prune `tt-metal-l2-nightly`

### PR DESCRIPTION
**Pull request by nostojicTT**
### Issues
https://github.com/tenstorrent/tt-metal/issues/53390

### PR Category
CI test pruning

### Summary
Drops `deepseek_prefill` nightly cases that exercise configurations production never runs. The filtering is done in Python at collection time, so the cases are *uncollected* rather than skipped, and `tests/pipeline_reorg/ops_unit_tests.yaml` keeps its baseline `-k` filters instead of growing a long deselect expression.

**≈55 min of L2-nightly wall clock saved: 3:02:15 → 2:07:42 (−29.9%)** across the seven `ttnn nightly experimental` jobs — or **−23.2% (≈34 min)** on the conservative read that drops the one noisy SKU. Collected tests fall 534 → 302 on both Blackhole SKUs. Per-job breakdown, run links and caveats under **Runtime impact** below.

**What is dropped — in CI only; a local run still collects everything:**

| Suite | Dropped | Rationale |
| --- | --- | --- |
| `test_deepseek_prefill_insert.py` | whole suite | op has no production counterpart |
| `test_deepseek_prefill_extract.py` | whole suite | op has no production counterpart |
| `test_single_routed_expert.py` | `x_tile` in both ISL sweeps | the routed expert is always fed a row-major activation |
| `test_moe_grouped_topk.py` | `in_fp32`, Blackhole | fp32 logits cover the pre-#51009 host-typecast gate contract; production feeds bf16 directly |
| `test_deepseek_prefill_per_token_cast.py` | `ROW_MAJOR_LAYOUT` input, Blackhole (4 test functions) | BH prefill always casts a tile-layout input |
| `test_deepseek_prefill_per_token_cast.py` | `scales_narrow_to_bf16`, Blackhole | BH keeps the scales at fp32 |
| `test_deepseek_prefill_per_token_cast.py` | precomputed scales, Blackhole | BH reads the scales back from the packed metadata |

**How it works — two new files, both under the `deepseek_prefill` test directory:**

- `conftest.py` — registers and applies an `uncollect_if(pred=...)` marker in `pytest_collection_modifyitems`. The predicate receives the case's collection-time params plus `is_ci_env` / `is_ci_v2_env` / `is_bh`; dropped items are reported through `pytest_deselected`. Same marker contract as `models/demos/deepseek_v3_d_p/tests/conftest.py`.
- `ci_pruning.py` — the six pruning rules, each defined once and shared by every decorator that needs it.

**Resulting counts in the `ttnn nightly experimental` job:**

| SKU | before | after |
| --- | --- | --- |
| `wh_n150_civ2` | 265 | 161 |
| `wh_n300_civ2` | 131 | 91 |
| `bh_p100` | 534 | 302 |
| `bh_p150b_civ2` | 534 | 302 |

Per-test before/after listing: https://claude.ai/code/artifact/92a0f490-0b04-4535-85c7-878e380aefff

### Runtime impact

**≈55 min saved across the seven `ttnn nightly experimental` jobs: 3:02:15 → 2:07:42 (−29.9%)**, matching the collected-count reduction above.

Job wall clock (includes per-job setup, unlike the test-step figures used in #53530 / #54314):

| Job | Before (avg of 2) | After | Delta |
| --- | ---: | ---: | ---: |
| `experimental (wormhole) [wh_n150_civ2]` | 48:38 | 37:30 | −22.9% |
| `experimental (wormhole) [wh_n300_civ2]` | 45:16 | 30:20 | −33.0% |
| `experimental (blackhole) [bh_p100]` | 37:11 | 16:16 | −56.3% |
| `experimental (blackhole) [bh_p150b_civ2]` | 36:14 | 28:27 | −21.5% |
| `experimental (blackhole - viommu)` | 5:21 | 5:33 | +3.6% |
| `generalized MoE gate op [wh_n150_civ2]` | 5:10 | 5:02 | −2.7% |
| `generalized MoE gate op [bh_p150b_civ2]` | 4:23 | 4:34 | +4.2% |
| **Total** | **3:02:15** | **2:07:42** | **−29.9%** |

Before: main runs [32944286374](https://github.com/tenstorrent/tt-metal/actions/runs/32944286374) (`3900b5ea463`) and [32722618739](https://github.com/tenstorrent/tt-metal/actions/runs/32722618739) (`298d4ef6798`), both verified to predate the merge commit `ee660b0ccd6`.
After: [32954712849](https://github.com/tenstorrent/tt-metal/actions/runs/32954712849), this branch's own dispatch.

Two caveats on those numbers:

- **The "after" is the branch run, not post-merge main.** The `experimental` category is not in every scheduled nightly: of the eight most recent main runs, it executed in four — all of them *before* the merge — and in none of the three since. So a main-to-main comparison is not yet possible, and the pruning only pays off on nightlies that actually select this category.
- **`bh_p100` is noisy and carries the largest delta.** Its two pre-merge samples were 21:19 and 53:04 — a 2.5× spread — so the −56.3% on that row is not trustworthy on its own. Excluding it, the remaining six jobs give **2:25:04 → 1:51:26, −23.2%** (≈34 min), which is the conservative read.

The collected-count reduction in the table above is the stronger evidence: it is deterministic, independent of machine load, and shows 534 → 302 on both Blackhole SKUs.

### Validation
`tt-metal-l2-nightly` dispatched on this branch with `additional_test_categories=experimental` (both archs, all other categories off), which runs the `ttnn nightly experimental` entries that own these suites: https://github.com/tenstorrent/tt-metal/actions/runs/32954712849

### Notes for reviewers
The pruning rules are the reviewable part — every entry in the table above is a claim that the configuration is not used in production. Predicate selection was verified on Blackhole to be id-identical to the equivalent `-k` selection.

### CODE REVIEW

**Review by kgrujcic-tt — APPROVED** [review]

File: tests/ttnn/nightly/unit_tests/operations/experimental/deepseek_prefill/conftest.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        if marker is None:
+            kept.append(item)
+            continue
+        params = dict(getattr(getattr(item, "callspec", None), "params", {}))
+        params.setdefault("is_ci_env", is_ci_env)
+        params.setdefault("is_ci_v2_env", is_ci_v2_env)
+        params.setdefault("is_bh", is_bh)
+        if not marker.kwargs["pred"](**params):
+            kept.append(item)
+    items[:] = kept
```
[the remark this answers is not in the document]
Comment (kgrujcic-tt) [reply tests/ttnn/nightly/unit_tests/operations/experimental/deepseek_prefill/conftest.py:40]: Done in 1c8fb823e05 — `pytest_collection_modifyitems` now collects the pruned items and passes them to `config.hook.pytest_deselected` before trimming `items`.

File: tests/ttnn/nightly/unit_tests/operations/experimental/deepseek_prefill/test_deepseek_prefill_per_token_cast.py [hunk 2]
```diff
@@ -130,6 +144,20 @@ def test_cast_to_fp8_scale(device, dtype, input_layout):
     assert_equal(scale, ref)
 
 
+def _ci_unsupported_param_combos_cast_to_fp8_scale_values(**params):
+    is_ci_env = params["is_ci_env"]
+    is_ci_v2_env = params["is_ci_v2_env"]
+    is_bh = params["is_bh"]
+    input_layout = params["input_layout"]
```
[the remark this answers is not in the document]
Comment (kgrujcic-tt) [reply tests/ttnn/nightly/unit_tests/operations/experimental/deepseek_prefill/test_deepseek_prefill_per_token_cast.py]: Done in 1c8fb823e05 — the rule is now `ci_pruning.bh_row_major_input`, defined once and referenced by all four decorators.

File: tests/ttnn/nightly/unit_tests/operations/experimental/deepseek_prefill/test_single_routed_expert.py [hunk 3]
```diff
@@ -326,6 +339,19 @@ def test_single_routed_expert_functional(
     )
 
 
+def _ci_unsupported_param_combos_isl_sweep(**params):
+    is_ci_env = params["is_ci_env"]
+    is_ci_v2_env = params["is_ci_v2_env"]
+    x_row_major = params["x_row_major"]
+
```
[the remark this answers is not in the document]
Comment (kgrujcic-tt) [reply tests/ttnn/nightly/unit_tests/operations/experimental/deepseek_prefill/test_single_routed_expert.py]: Done in 1c8fb823e05 — both sweeps now use the single sweep-neutral `ci_pruning.tiled_x_input`.

File: tests/ttnn/nightly/unit_tests/operations/experimental/deepseek_prefill/test_deepseek_prefill_extract.py [hunk 4]
```diff
@@ -17,6 +17,18 @@
 from tests.ttnn.utils_for_testing import assert_with_pcc
 
 
+def _ci_unsupported_param_combos(**params):
+    is_ci_env = params["is_ci_env"]
+    is_ci_v2_env = params["is_ci_v2_env"]
+
+    if not (is_ci_env or is_ci_v2_env):
```
[the remark this answers is not in the document]
Comment (kgrujcic-tt) [reply tests/ttnn/nightly/unit_tests/operations/experimental/deepseek_prefill/test_deepseek_prefill_extract.py]: Done in 1c8fb823e05 — insert and extract now share `ci_pruning.no_production_counterpart`, defined in `ci_pruning.py` next to the collection hook.
